server/controllers/shared/sharedController.ts

```diff
@@ -30,31 +35,70 @@ export default class SharedController {
     required: [],
   }
 
-  constructor(private reportService: ReportService = null) {}
+  updateReport: () => void
+
+  constructor(protected readonly reportService: ReportService = null, protected report: Report = null) {}
 
   protected renderTemplate(res: Response, templateValues: TemplateValues) {
     res.render(`${this.path}/${this.templatePath}`, templateValues)
   }
 
-  get = async (req: Request, res: Response): Promise<void> => {
-    const report = await this.reportService.getReportById(req.params.reportId)
-    if (report) {
+  private getStoredData = () => {
+    this.data = {}
+    if (this.report && this.report.fieldValues) {
+      this.report.fieldValues.forEach(item => {
+        if (this.pageFields.includes(item.field.name)) {
+          this.data[item.field.name] = item.value
+        }
+      })
+    }
+  }
+
+  protected updateFields = async (formData: unknown) => {
+    const fieldValues: Array<IFieldValue> = []
+    if (this.report && this.report.reportDefinition && this.report.reportDefinition.fields) {
+      this.report.reportDefinition.fields.forEach(item => {
+        if (formData[item.name]) {
+          const fieldValue = this.report.fieldValues.filter(value => item.name === value.field.name).pop()
+          fieldValues.push({
+            reportId: this.report.id,
+            fieldId: item.id,
+            value: Array.isArray(formData[item.name]) ? formData[item.name].join(',') : formData[item.name],
+            version: (fieldValue && fieldValue.version) || 1,
+          })
+        }
+      })
+    }
+    if (fieldValues.length) {
+      await this.reportService.updateFieldValues(fieldValues)
+    }
+  }
+
+  public get = async (req: Request, res: Response): Promise<void> => {
+    this.report = await this.reportService.getReportById(req.params.reportId)
+    if (this.report) {
+      this.getStoredData()
+      if (this.updateReport) {
+        this.updateReport()
```

Comment:
Does this ever get set other than on line 38?

Reply:
Yes, this is overridden in the parent class where necessary.